src/libpgmoneta/se_ssh.c

```diff
@@ -0,0 +1,468 @@
+         goto error;
+   }
+
+   key_path = malloc(strlen(homedir) + strlen(pubkey_path) + 2);
+   memset(key_path, 0, strlen(homedir) + strlen(pubkey_path) + 2);
+   strcpy(key_path, homedir);
+   strcat(key_path, pubkey_path);
+
+   rc = ssh_pki_import_pubkey_file(key_path, &client_pubkey);
+   if (rc != SSH_OK)
+   {
+      goto error;
+   }
+
+   memset(key_path, 0, strlen(homedir) + strlen(pubkey_path) + 2);
+   strcpy(key_path, homedir);
+   strcat(key_path, privkey_path);
+
+   rc = ssh_pki_import_privkey_file(key_path, NULL, NULL, NULL,
+                                    &client_privkey);
+   if (rc != SSH_OK)
+   {
+      goto error;
+   }
+
+   rc = ssh_userauth_publickey(session, NULL, client_privkey);
+   if (rc != SSH_AUTH_SUCCESS)
+   {
+      printf("%d", rc);
+      goto error;
+   }
+
+   sftp = sftp_new(session);
+
+   if (sftp == NULL)
+   {
+      pgmoneta_log_error("Error: %s\n", ssh_get_error(session));
+      goto error;
+   }
+
+   rc = sftp_init(sftp);
+   if (rc != SSH_OK)
+   {
+      pgmoneta_log_error("Error: %s\n", sftp_get_error(sftp));
+      goto error;
+   }
+
+   ssh_string_free_char(hexa);
+   ssh_clean_pubkey_hash(&srv_pubkey_hash);
+   ssh_key_free(srv_pubkey);
+   ssh_key_free(client_pubkey);
+   ssh_key_free(client_privkey);
+
+   free(key_path);
+
+   return 0;
+
+error:
+   ssh_string_free_char(hexa);
+   ssh_clean_pubkey_hash(&srv_pubkey_hash);
+   ssh_key_free(srv_pubkey);
+   ssh_key_free(client_pubkey);
+   ssh_key_free(client_privkey);
+
+   free(key_path);
+
+   sftp_free(sftp);
+
+   ssh_disconnect(session);
+   ssh_free(session);
+   return 1;
+}
+
+static int
+ssh_storage_execute(int server, char* identifier,
+                    struct node* i_nodes, struct node** o_nodes)
+{
+   char* local_root = NULL;
+   char* remote_root = NULL;
+   struct configuration* config;
+
+   config = (struct configuration*)shmem;
+
+   remote_root = pgmoneta_append(remote_root, config->ssh_base_dir);
+   if (!pgmoneta_ends_with(config->ssh_base_dir, "/"))
+   {
+      remote_root = pgmoneta_append(remote_root, "/");
+   }
+
+   remote_root = pgmoneta_append(remote_root, config->servers[server].name);
+   remote_root = pgmoneta_append(remote_root, "/backup/");
+   remote_root = pgmoneta_append(remote_root, identifier);
+
+   if (pgmoneta_sftp_mkdir(remote_root) == 1)
+   {
+      goto error;
+   }
+
+   local_root = pgmoneta_get_server_backup_identifier(server, identifier);
+
+   if (pgmoneta_sftp_copy_directory(local_root, remote_root) != 0)
+   {
+      goto error;
+   }
+
+   free(remote_root);
+
+   free(local_root);
+
+   return 0;
+
+error:
+
+   free(remote_root);
+
+   free(local_root);
+
+   return 1;
+}
+
+static int
+ssh_storage_teardown(int server, char* identifier,
+                     struct node* i_nodes, struct node** o_nodes)
+{
+   char* root = NULL;
+
+   root = pgmoneta_get_server_backup_identifier_data(server, identifier);
+   pgmoneta_delete_directory(root);
```

Comment:
We need to explain when `local` is deleted

... there is the question of the metadata used for commands like `details` ...